src/util.rs

```diff
@@ -67,3 +67,70 @@ where
         P::COLOR_TYPE,
     )
 }
+
+mod parse_geometry {
+    use crate::util;
+
+    use nom::bytes::complete as bytes;
+    use nom::character::complete as chr;
+    use nom::combinator as comb;
+    use nom::sequence as seq;
+
+    fn equal_sign(i: &str) -> nom::IResult<&str, char> {
+        chr::char('=')(i)
+    }
+
+    fn x_sign(i: &str) -> nom::IResult<&str, char> {
+        chr::one_of("xX")(i)
+    }
+
+    fn integer(i: &str) -> nom::IResult<&str, i32> {
+        comb::map_res(
+            // Limit to 5 digits - X11 uses i16 and u16 for position and size,
+            // and this fits comfortably into our i32 Rects.
+            bytes::take_while_m_n(1, 5, |c: char| c.is_ascii_digit()),
+            str::parse,
+        )(i)
+    }
+
+    fn sign(i: &str) -> nom::IResult<&str, i32> {
+        comb::map(chr::one_of("-+"), |s| if s == '-' { -1 } else { 1 })(i)
+    }
+
+    fn signed_integer(i: &str) -> nom::IResult<&str, i32> {
+        comb::map(seq::pair(sign, integer), |(s, m)| s * m)(i)
+    }
+
+    /// Parse a string of the form `=<width>x<height>{+-}<xoffset>{+-}<yoffset>` into a [`util::Rect`].
+    pub fn parse_geometry(g: &str) -> Option<util::Rect> {
+        let (remainder, (_, w, _, h, x, y)) = seq::tuple((
+            comb::opt(equal_sign),
+            integer,
+            x_sign,
+            integer,
+            signed_integer,
+            signed_integer,
+        ))(g)
+        .ok()?;
+
+        if !remainder.is_empty() {
```

Comment:
Negation here seems a bit weird.

Reply:
It's really meant to be a guard clause: if we have any remaining bytes in the input, then it is invalid therefore we reject it. I suppose the `else` is the weird part here. This could be:

```rust
if !remainder.is_empty() {
    return None;
}

Some(...)
```

Reply:
It was more about swapping the branches and removing the negation to make it slightly easier to read. I think I was tired and took this for a double negation, but it's only one, so it's fine really.